File: paging/common/src/main/java/androidx/paging/Pager.java
/*
 * Copyright 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.paging;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.arch.core.util.Function;
import androidx.paging.futures.FutureCallback;
import androidx.paging.futures.Futures;

import com.google.common.util.concurrent.ListenableFuture;

import java.util.concurrent.Executor;
import java.util.concurrent.atomic.AtomicBoolean;

public class Pager<K, V> {

    static <K, V> ListenableFuture<Pager<K, V>> createPager(
            @NonNull final ListenableSource.LoadType type,
            @Nullable final K initialKey,
            @NonNull final PagedList.Config config,
            @NonNull final ListenableSource<K, V> source,
            @NonNull final Executor notifyExecutor,
            @NonNull final AdjacentProvider<V> adjacentProvider,
            @NonNull final PageConsumer<V> pageConsumer
    ) {
        return Futures.transform(
                source.load(new ListenableSource.Params<>(
                        type,
                        initialKey,
                        config.initialLoadSizeHint,
                        config.enablePlaceholders,
                        config.pageSize)),
                new Function<ListenableSource.BaseResult<V>, Pager<K, V>>() {
                    @Override
                    public Pager<K, V> apply(ListenableSource.BaseResult<V> result) {
                        return new Pager<>(config, source, notifyExecutor, pageConsumer,
                                adjacentProvider, result);
                    }
                }, notifyExecutor);
    }

    @SuppressWarnings({"WeakerAccess", "unchecked"}) /* synthetic accessor */
    Pager(@NonNull PagedList.Config config,
            @NonNull ListenableSource<K, V> source,
            @NonNull Executor notifyExecutor,
            @NonNull PageConsumer<V> pageConsumer,
            @Nullable AdjacentProvider<V> adjacentProvider,
            @NonNull ListenableSource.BaseResult<V> result) {
        mConfig = config;
        mSource = source;
        mNotifyExecutor = notifyExecutor;
        mPageConsumer = pageConsumer;
        if (adjacentProvider == null) {
            adjacentProvider = new SimpleAdjacentProvider<>();
        }
        mAdjacentProvider = adjacentProvider;
        mPrevKey = (K) result.mPrevKey;
        mNextKey = (K) result.mNextKey;
    }

    @NonNull
    private final PagedList.Config mConfig;
    @NonNull
    private final ListenableSource<K, V> mSource;
    @NonNull
    private final Executor mNotifyExecutor;
    @NonNull
    private final PageConsumer<V> mPageConsumer;
    @NonNull
    private final AdjacentProvider<V> mAdjacentProvider;
    @Nullable
    final private K mPrevKey;
    @Nullable
    final private K mNextKey;

    private final AtomicBoolean mDetached = new AtomicBoolean(false);

    PagedList.LoadStateManager mLoadStateManager = new PagedList.LoadStateManager() {
        @Override
        protected void onStateChanged(@NonNull PagedList.LoadType type,
                @NonNull PagedList.LoadState state, @Nullable Throwable error) {
            System.out.println("onStateChanged " + type + ", state " + state);
            mPageConsumer.onStateChanged(type, state, error);
        }
    };

    private void listenTo(@NonNull final PagedList.LoadType type,
            @NonNull ListenableFuture<? extends ListenableSource.BaseResult<V>> future) {
        // TODO: check if is invalid on BG executor, if it is, detach()
        Futures.addCallback(future, new FutureCallback<ListenableSource.BaseResult<V>>() {
            @Override
            public void onSuccess(ListenableSource.BaseResult<V> value) {
                onLoadSuccess(type, value);
            }

            @Override
            public void onError(@NonNull Throwable throwable) {
                onLoadError(type, throwable);
            }
        }, mNotifyExecutor);
    }

    interface PageConsumer<V> {
        // return true if we need to fetch more
        boolean onPageResult(
                @NonNull PagedList.LoadType type,
                @NonNull ListenableSource.BaseResult<V> pageResult);

        void onStateChanged(@NonNull PagedList.LoadType type,
                @NonNull PagedList.LoadState state, @Nullable Throwable error);
    }

    interface AdjacentProvider<V> {
        V getFirstLoadedItem();

        V getLastLoadedItem();

        int getFirstLoadedItemIndex();

        int getLastLoadedItemIndex();

        // TODO: use this for non-contiguouspagedlist usecase
        void onPageResultResolution(
                @NonNull PagedList.LoadType type,
                @NonNull PageResult<V> pageResult);
    }

    private void onLoadSuccess(PagedList.LoadType type, ListenableSource.BaseResult<V> value) {
        if (isDetached()) {
            // abort!
            return;
        }
        if (mPageConsumer.onPageResult(type, value)) {
            if (type.equals(PagedList.LoadType.START)) {
                schedulePrepend();
            } else if (type.equals(PagedList.LoadType.END)) {
                scheduleAppend();
            } else {
                throw new IllegalStateException("Can only fetch more during append/prepend");
            }
        } else {
            PagedList.LoadState state = value.mData.isEmpty() ? PagedList.LoadState.DONE : PagedList.LoadState.IDLE;
            mLoadStateManager.setState(type, state, null);
        }
    }

    private void onLoadError(PagedList.LoadType type, Throwable throwable) {
        if (isDetached()) {
            // abort!
            return;
        }
        // TODO: handle nesting
        PagedList.LoadState state = mSource.isRetryableError(throwable)
                ? PagedList.LoadState.RETRYABLE_ERROR : PagedList.LoadState.ERROR;
        mLoadStateManager.setState(type, state, throwable);
    }

    public void trySchedulePrepend() {
        if (mLoadStateManager.getStart().equals(PagedList.LoadState.IDLE)) {
            schedulePrepend();
        }
    }

    public void tryScheduleAppend() {
        if (mLoadStateManager.getEnd().equals(PagedList.LoadState.IDLE)) {
            scheduleAppend();
        }
    }

    private void schedulePrepend() {

        K key;
        switch(mSource.mType) {
            case POSITIONAL:
                //noinspection unchecked
                key = (K) ((Integer) mAdjacentProvider.getFirstLoadedItemIndex());
                break;
            case PAGE_KEYED:
                key = mNextKey;
                break;
            case ITEM_KEYED:
                key = mSource.getKey(mAdjacentProvider.getFirstLoadedItem());
                break;
            default:
                throw new IllegalArgumentException("unknown source type");
        }
        mLoadStateManager.setState(PagedList.LoadType.START, PagedList.LoadState.LOADING,null);
        listenTo(PagedList.LoadType.START, mSource.load(new ListenableSource.Params<>(
                ListenableSource.LoadType.START,
                key,
                mConfig.initialLoadSizeHint,
                mConfig.enablePlaceholders,
                mConfig.pageSize)));
    }

    private void scheduleAppend() {
        K key;
        switch(mSource.mType) {
            case POSITIONAL:
                //noinspection unchecked
                key = (K) ((Integer) mAdjacentProvider.getLastLoadedItemIndex());
                System.out.println("scheduleAppend, new last " + key);
                break;
            case PAGE_KEYED:
                key = mNextKey;
                break;
            case ITEM_KEYED:
                key = mSource.getKey(mAdjacentProvider.getLastLoadedItem());
                break;
            default:
                throw new IllegalArgumentException("unknown source type");
        }
        System.out.println("schedule append, loading...");
        mLoadStateManager.setState(PagedList.LoadType.END, PagedList.LoadState.LOADING,null);
        listenTo(PagedList.LoadType.END, mSource.load(new ListenableSource.Params<>(
                ListenableSource.LoadType.END,
                key,
                mConfig.initialLoadSizeHint,
                mConfig.enablePlaceholders,
                mConfig.pageSize)));
    }

    void retry() {
        if (mLoadStateManager.getStart().equals(PagedList.LoadState.RETRYABLE_ERROR)) {
            schedulePrepend();
        }
        if (mLoadStateManager.getEnd().equals(PagedList.LoadState.RETRYABLE_ERROR)) {
            scheduleAppend();
        }
    }


    public boolean isDetached() {
        return mDetached.get();
    }

    public void detach() {
        mDetached.set(true);
    }

    static class SimpleAdjacentProvider<V> implements AdjacentProvider<V> {
        private int mFirstIndex;
        private int mLastIndex;

        private V mFirstItem;
        private V mLastItem;

        @Override
        public V getFirstLoadedItem() {
            return mFirstItem;
        }

        @Override
        public V getLastLoadedItem() {
            return mLastItem;
        }

        @Override
        public int getFirstLoadedItemIndex() {
            return mFirstIndex;
        }

        @Override
        public int getLastLoadedItemIndex() {
            return mLastIndex;
        }

        @Override
        public void onPageResultResolution(@NonNull PagedList.LoadType type, @NonNull PageResult<V> pageResult) {
            if (type == PagedList.LoadType.START) {
                mFirstIndex -= pageResult.page.size();
                mFirstItem = pageResult.page.get(0);
            } else if (type == PagedList.LoadType.END) {
                mLastIndex += pageResult.page.size();
                mLastItem = pageResult.page.get(pageResult.page.size() - 1);
            } else {
                mFirstIndex = pageResult.leadingNulls;
                mLastIndex = pageResult.leadingNulls + pageResult.page.size();
                mFirstItem = pageResult.page.get(0);
                mLastItem = pageResult.page.get(pageResult.page.size() - 1);
            }
        }
    }
}

File: paging/common/src/main/kotlin/androidx/paging/Pager.kt
/*
 * Copyright 2019 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.paging

import androidx.paging.futures.FutureCallback
import androidx.paging.futures.Futures

import com.google.common.util.concurrent.ListenableFuture

import java.util.concurrent.Executor
import java.util.concurrent.atomic.AtomicBoolean

internal class Pager<K, V>(
    val config: PagedList.Config/* synthetic access */,
    val source: DataSource<K, V>/* synthetic access */,
    val mNotifyExecutor: Executor/* synthetic access */,
    private val mFetchExecutor: Executor,
    val mPageConsumer: PageConsumer<V>/* synthetic access */,
    adjacentProvider: AdjacentProvider<V>?,
    result: DataSource.BaseResult<V>
) {

    private val mTotalCount: Int

    private val adjacentProvider: AdjacentProvider<V>

    private var mPrevKey: K? = null

    private var nextKey: K? = null

    private val mDetached = AtomicBoolean(false)

    var loadStateManager: PagedList.LoadStateManager = object : PagedList.LoadStateManager() {
        override fun onStateChanged(
            type: PagedList.LoadType,
            state: PagedList.LoadState,
            error: Throwable?
        ) {
            mPageConsumer.onStateChanged(type, state, error)
        }
    }

    val isDetached: Boolean
        get() = mDetached.get()

    init {
        this.adjacentProvider = adjacentProvider ?: SimpleAdjacentProvider()
        @Suppress("UNCHECKED_CAST")
        mPrevKey = result.prevKey as K?
        @Suppress("UNCHECKED_CAST")
        nextKey = result.nextKey as K?
        this.adjacentProvider.onPageResultResolution(PagedList.LoadType.REFRESH, result)
        mTotalCount = result.totalCount()

        // TODO: move this validation to tiled paging impl, once that's added back
        if (source.type === DataSource.KeyType.POSITIONAL && config.enablePlaceholders) {
            result.validateForInitialTiling(config.pageSize)
        }
    }

    private fun listenTo(
        type: PagedList.LoadType,
        future: ListenableFuture<out DataSource.BaseResult<V>>
    ) {
        // First listen on the BG thread if the DataSource is invalid, since it can be expensive
        future.addListener(Runnable {
            // if invalid, drop result on the floor
            if (source.isInvalid()) {
                detach()
                return@Runnable
            }

            // Source has been verified to be valid after producing data, so sent data to UI
            Futures.addCallback(
                future,
                object : FutureCallback<DataSource.BaseResult<V>> {
                    override fun onSuccess(value: DataSource.BaseResult<V>) {
                        onLoadSuccess(type, value)
                    }

                    override fun onError(throwable: Throwable) {
                        onLoadError(type, throwable)
                    }
                },
                mNotifyExecutor
            )
        }, mFetchExecutor)
    }

    internal interface PageConsumer<V> {
        // return true if we need to fetch more
        fun onPageResult(
            type: PagedList.LoadType,
            pageResult: DataSource.BaseResult<V>
        ): Boolean

        fun onStateChanged(
            type: PagedList.LoadType,
            state: PagedList.LoadState,
            error: Throwable?
        )
    }

    internal interface AdjacentProvider<V> {
        val firstLoadedItem: V?

        val lastLoadedItem: V?

        val firstLoadedItemIndex: Int

        val lastLoadedItemIndex: Int

        /**
         * Notify the AdjacentProvider of new loaded data, to update first/last item/index.
         *
         * NOTE: this data may not be committed (e.g. it may be dropped due to max size). Up to the
         * implementation of the AdjacentProvider to handle this (generally by ignoring this
         * call if dropping is supported).
         */
        fun onPageResultResolution(
            type: PagedList.LoadType,
            result: DataSource.BaseResult<V>
        )
    }

    fun onLoadSuccess(type: PagedList.LoadType, value: DataSource.BaseResult<V>) {
        if (isDetached) {
            // abort!
            return
        }

        adjacentProvider.onPageResultResolution(type, value)

        if (mPageConsumer.onPageResult(type, value)) {
            if (type == PagedList.LoadType.START) {
                @Suppress("UNCHECKED_CAST")
                mPrevKey = value.prevKey as K?
                schedulePrepend()
            } else if (type == PagedList.LoadType.END) {
                @Suppress("UNCHECKED_CAST")
                nextKey = value.nextKey as K?
                scheduleAppend()
            } else {
                throw IllegalStateException("Can only fetch more during append/prepend")
            }
        } else {
            val state =
                if (value.data.isEmpty()) PagedList.LoadState.DONE else PagedList.LoadState.IDLE
            loadStateManager.setState(type, state, null)
        }
    } /* synthetic accessor */

    fun onLoadError(type: PagedList.LoadType, throwable: Throwable) {
        if (isDetached) {
            // abort!
            return
        }
        // TODO: handle nesting
        val state = if (source.isRetryableError(throwable))
            PagedList.LoadState.RETRYABLE_ERROR
        else
            PagedList.LoadState.ERROR
        loadStateManager.setState(type, state, throwable)
    } /* synthetic accessor */

    fun trySchedulePrepend() {
        if (loadStateManager.start == PagedList.LoadState.IDLE) {
            schedulePrepend()
        }
    }

    fun tryScheduleAppend() {
        if (loadStateManager.end == PagedList.LoadState.IDLE) {
            scheduleAppend()
        }
    }

    private fun canPrepend(): Boolean {
        return if (mTotalCount == DataSource.BaseResult.TOTAL_COUNT_UNKNOWN) {
            // don't know count / position from initial load, so be conservative, return true
            true
        } else adjacentProvider.firstLoadedItemIndex > 0

        // position is known, do we have space left?
    }

    private fun canAppend(): Boolean {
        return if (mTotalCount == DataSource.BaseResult.TOTAL_COUNT_UNKNOWN) {
            // don't know count / position from initial load, so be conservative, return true
            true
        } else adjacentProvider.lastLoadedItemIndex < mTotalCount - 1

        // count is known, do we have space left?
    }

    private fun schedulePrepend() {
        if (!canPrepend()) {
            onLoadSuccess(PagedList.LoadType.START, DataSource.BaseResult.empty())
            return
        }

        @Suppress("UNCHECKED_CAST")
        val key = when (source.type) {
            DataSource.KeyType.POSITIONAL ->
                (adjacentProvider.firstLoadedItemIndex - 1) as K
            DataSource.KeyType.PAGE_KEYED -> mPrevKey
            DataSource.KeyType.ITEM_KEYED -> (source as ListenableItemKeyedDataSource).getKey(
                adjacentProvider.firstLoadedItem!!
            )
        }

        loadStateManager.setState(PagedList.LoadType.START, PagedList.LoadState.LOADING, null)
        listenTo(
            PagedList.LoadType.START, source.load(
                DataSource.Params(
                    DataSource.LoadType.START,
                    key,
                    config.initialLoadSizeHint,
                    config.enablePlaceholders,
                    config.pageSize
                )
            )
        )
    }

    private fun scheduleAppend() {
        if (!canAppend()) {
            onLoadSuccess(PagedList.LoadType.END, DataSource.BaseResult.empty())
            return
        }

        @Suppress("UNCHECKED_CAST")
        val key = when (source.type) {
            DataSource.KeyType.POSITIONAL -> (adjacentProvider.lastLoadedItemIndex + 1) as K
            DataSource.KeyType.PAGE_KEYED -> nextKey
            DataSource.KeyType.ITEM_KEYED -> (source as ListenableItemKeyedDataSource).getKey(
                adjacentProvider.lastLoadedItem!!
            )
        }

        loadStateManager.setState(PagedList.LoadType.END, PagedList.LoadState.LOADING, null)
        listenTo(
            PagedList.LoadType.END, source.load(
                DataSource.Params(
                    DataSource.LoadType.END,
                    key,
                    config.initialLoadSizeHint,
                    config.enablePlaceholders,
                    config.pageSize
                )
            )
        )
    }

    fun retry() {
        if (loadStateManager.start == PagedList.LoadState.RETRYABLE_ERROR) {
            schedulePrepend()
        }
        if (loadStateManager.end == PagedList.LoadState.RETRYABLE_ERROR) {
            scheduleAppend()
        }
    }

    fun detach() {
        mDetached.set(true)
    }

    internal class SimpleAdjacentProvider<V> : AdjacentProvider<V> {
        override var firstLoadedItemIndex: Int = 0
            private set
        override var lastLoadedItemIndex: Int = 0
            private set

        override var firstLoadedItem: V? = null
            private set
        override var lastLoadedItem: V? = null
            private set

        var counted: Boolean = false
        var leadingUnloadedCount: Int = 0
        var trailingUnloadedCount: Int = 0

        override fun onPageResultResolution(
            type: PagedList.LoadType,
            result: DataSource.BaseResult<V>
        ) {
            if (result.data.isEmpty()) {
                return
            }
            if (type == PagedList.LoadType.START) {
                firstLoadedItemIndex -= result.data.size
                firstLoadedItem = result.data[0]
                if (counted) {
                    leadingUnloadedCount -= result.data.size
                }
            } else if (type == PagedList.LoadType.END) {
                lastLoadedItemIndex += result.data.size
                lastLoadedItem = result.data.last()
                if (counted) {
                    trailingUnloadedCount -= result.data.size
                }
            } else {
                firstLoadedItemIndex = result.leadingNulls + result.offset
                lastLoadedItemIndex = firstLoadedItemIndex + result.data.size - 1
                firstLoadedItem = result.data[0]
                lastLoadedItem = result.data.last()

                if (result.counted) {
                    counted = true
                    leadingUnloadedCount = result.leadingNulls
                    trailingUnloadedCount = result.trailingNulls
                }
            }
        }
    }
}
